src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -283,41 +320,74 @@ void BlockchainSQLite::upgrade_schema() {
+    // - delayed_payments_prune into `delayed_payments`
     {
-        SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
-        db.exec(R"(
+        auto& netconf = get_config(m_nettype);
+        db.exec(
+                R"(
+        -- Keep a copy of all the rows for earnt rewards for this height if it's on an archival
+        -- interval. It allows the DB to gracefully handle block re-orgs without having to
+        -- recalculate from scratch.
+        --
+        -- We archive state at every 'HISTORY_ARCHIVE_INTERVAL' height and we prune the stored
+        -- archive to encompass the past 'HISTORY_ARCHIVE_WINDOW' blocks worth of history.
+        --
+        -- When pruning we floor to the closest interval to make the equivalent pruning math in the
+        -- SNL at 'process_block()' simple which has additional constraints.
+        DROP TRIGGER IF EXISTS make_archive;
+        CREATE TRIGGER make_archive AFTER UPDATE ON batch_db_info
+        FOR EACH ROW WHEN (NEW.height % {}) = 0 AND NEW.height > OLD.height BEGIN
+            INSERT INTO batched_payments_accrued_archive SELECT *, NEW.height FROM batched_payments_accrued;
+            DELETE FROM batched_payments_accrued_archive WHERE height < ((NEW.height - {}) % {});
+        END;
+
+        -- On re-org to an older height delete all archive rows that are newer than the DB's height
         DROP TRIGGER IF EXISTS clear_archive;
         CREATE TRIGGER clear_archive AFTER UPDATE ON batch_db_info
         FOR EACH ROW WHEN NEW.height < OLD.height BEGIN
-            DELETE FROM batched_payments_accrued_archive WHERE archive_height > NEW.height;
+            DELETE FROM batched_payments_accrued_archive WHERE height > NEW.height;
         END;
 
+        -- On re-org delete all recent rows that are newer than the DB's height
         DROP TRIGGER IF EXISTS clear_recent;
         CREATE TRIGGER clear_recent AFTER UPDATE ON batch_db_info
         FOR EACH ROW WHEN NEW.height < OLD.height BEGIN
             DELETE FROM batched_payments_accrued_recent WHERE height > NEW.height;
         END;
-        )");
-        transaction.commit();
+
+        -- Saves the current accrued rewards table into the recent table for the current height
+        DROP TRIGGER IF EXISTS make_recent;
+        CREATE TRIGGER make_recent AFTER UPDATE ON batch_db_info
+        FOR EACH ROW WHEN NEW.height > OLD.height BEGIN
+            INSERT INTO batched_payments_accrued_recent SELECT *, NEW.height FROM batched_payments_accrued;
+            DELETE FROM batched_payments_accrued_recent WHERE height < NEW.height - {};
+        END;
+
+        -- Delete old delayed payments from the DB when they are processed
+        DROP TRIGGER IF EXISTS delayed_payments_prune;
+        CREATE TRIGGER delayed_payments_prune AFTER UPDATE ON batch_db_info
+        FOR EACH ROW BEGIN
+            DELETE FROM delayed_payments WHERE payout_height < NEW.height;
```

Comment:
I *think* this needs to be `WHERE entry_height < NEW.height`.  As-is, a re-org right after a delayed payment is paid out would revert the payout but no longer have the delayed payment entry in this table.

Reply:
So you've actually uncovered a bunch of bugs having reviewed this after your findings. We need 2 triggers here, one from blockchain detached and one for when a block is added.

This one you've pointed out was originally written to handle the `on_block_add` case. When we add a block, we process all the transactions in the block, give users back their stake (delayed payment) if it's queued for this block, then we execute this trigger which deletes these delayed payment that has been executed.

Then there's the situation where the blockchain reorgs/detaches. In that case we need the condition you've mentioned but I believe with `>` instead of `<` which is that `entry.height > NEW.height`. Then when oxen replays the chain it'll recreate the delayed payments when it passes at the `entry.height`.

So this stuff's pretty confusing, let me know if what I've summarised is correct. Also note that the SQL stores the chain height not the chain length, i.e. it stores the block number that it last executed so that's the reasoning for `<=` and `>` and so forth in the commit.

Reply:
> When we add a block, we process all the transactions in the block, give users back their stake (delayed payment) if it's queued for this block, then we execute this trigger which deletes these delayed payment that has been executed.

Since we're not snapshotting this table, the chain of events `add block -> payout -> delete row because payout made -> "pop" block (well, some small amount of detach)` will lead to that delayed payment not happening if the detach lands between `entry.height` and `payout.height`, no?  Which...hmm, we might actually need to snapshot it for that reason now that I think about it.  Like, if we want to be able to say "those DBs got out of sync, rewind them to synchronize" without going from the beginning of time, I'm pretty sure we can rewind to such a situation.

Reply:
Oh you're right. I wonder if this is the thing breaking the blockchain syncing when we sync dev onto stable because right now it does a big rescan from some earlier checkpoint, the old trigger was this

```
        CREATE TRIGGER delayed_payments_prune AFTER UPDATE ON batch_db_info
        FOR EACH ROW BEGIN
            DELETE FROM delayed_payments WHERE payout_height < (NEW.height - 10000);
        END;
```

I was like what's up with the magic number, that'd break in the way you describe if you go further back than 10k blocks. Well .. I didn't want to but I guess I'll have to do the same archive table thingy with the delayed payments.

Reply:
So I've gone and done this and I had to shuffle some more code around but I think it's in a better place now. I merged a bunch of triggers together, so there's one trigger for block-reorg, one trigger to save rows on block add and so forth. So it's less scattered all over the place.

This broke a couple of things so I had to go and debug them and issue some fixes where more bugs were found. I made the integration test more complicated to check for this `add block -> payout -> delete row because payout made -> "pop" block (well, some small amount of detach)` situation, took a bit of time due to how long it takes to run those tests, iterate and verify.